[StructMech] Displacement Control Introduction

zidanmahmoud opened the pull request:
Currently only supports one node in one global direction.

Commit:
adding displacement control support

Commit:
bug fix

loumalouomega commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -0,0 +1,345 @@
+
+    for (SizeType i = 0; i < number_of_nodes; ++i) {
+        int index = i * 2;
+        if (GetValue(POINT_LOAD_X) != 0.0) // prescribed displacement in x direction
```
I would not use this comparison. I would access to the whole POINT_LOAD vector and then check component by component using the epsilon value

loumalouomega commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -0,0 +1,345 @@
+
+    for (SizeType i = 0; i < number_of_nodes; ++i) {
+        SizeType index = i * block_size;
+        if (GetValue(POINT_LOAD_X) != 0.0) // prescribed displacement in x direction
```
Idem, I would create a private method in order to avoid duplicated code

loumalouomega commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -0,0 +1,345 @@
+
+    for (SizeType i = 0; i < number_of_nodes; ++i) {
+        SizeType index = i * GetBlockSize();
+        if (GetValue(POINT_LOAD_X) != 0.0) // prescribed displacement in x direction
```
Idem, private method

loumalouomega commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -0,0 +1,345 @@
+            rLeftHandSideMatrix.resize(mat_size, mat_size, false);
+        }
+        noalias(rLeftHandSideMatrix) = ZeroMatrix(mat_size, mat_size);
+        if (GetValue(POINT_LOAD_X) != 0.0) // prescribed displacement in x direction
```
Idem private method

loumalouomega commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -0,0 +1,345 @@
+
+        rRightHandSideVector(0) += GetGeometry()[0].FastGetSolutionStepValue(LOAD_FACTOR);
+        rRightHandSideVector(1) += GetValue(PRESCRIBED_DISPLACEMENT);
+        if (GetValue(POINT_LOAD_X) != 0.0) // prescribed displacement in x direction
```
Idem private method

loumalouomega reviewed (changes requested):
See my comments. Add a test please. Good job BTW

philbucher commented:
cool!

as @loumalouomega said please add a test. Before you do so please talk to us how to make a suitable test for this

Commit:
adding test for displacement control

Commit:
better direction check

zidanmahmoud commented:
I added a test with a Two-Bar truss problem as in the [wiki](https://github.com/KratosMultiphysics/Kratos/wiki/Two-bar-truss---snap-through).
Also added two private methods for checking direction, one for point loads and one for displacement.

zidanmahmoud commented:
Also ran a more complex case, the arc of trusses
![Snapshot](https://user-images.githubusercontent.com/43052032/67112786-54d53680-f1d8-11e9-8214-2213c5e65a8e.png)

and the results are correct
![arc](https://user-images.githubusercontent.com/43052032/67112854-746c5f00-f1d8-11e9-8305-7e7dfdeda5b5.png)

loumalouomega commented:
That example can go to the Example repository

El vie., 18 oct. 2019 19:37, Mahmoud Zidan <notifications@github.com>
escribió:

> Also ran a more complex case, the arc of trusses
> [image: Snapshot]
> <https://user-images.githubusercontent.com/43052032/67112786-54d53680-f1d8-11e9-8214-2213c5e65a8e.png>
>
> and the results are correct
> [image: arc]
> <https://user-images.githubusercontent.com/43052032/67112854-746c5f00-f1d8-11e9-8305-7e7dfdeda5b5.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/5799?email_source=notifications&email_token=AEYQZAB4PJPXCT2FCMO6FBDQPHS5HA5CNFSM4JB45VKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBVECSA#issuecomment-543834440>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEYQZAEEQBVV2TNOYDOMUQ3QPHS5HANCNFSM4JB45VKA>
> .
>

loumalouomega commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -104,6 +104,45 @@ Condition::Pointer DisplacementControlCondition::Clone (
+    if( this->Has( POINT_LOAD ) )
+        noalias(point_load) = this->GetValue( POINT_LOAD );
+
+    if (abs(point_load(0)) > numerical_limit) // prescribed displacement in x direction
```
NOOOOOOO
![imagen](https://user-images.githubusercontent.com/19991680/35767644-8a9c6856-08f0-11e8-9944-e72c6c1db7b3.png)

abs, instead of std::abs, returns an integer

zidanmahmoud replied:
Oh :man_facepalming: sorry

loumalouomega reviewed (changes requested):
std::abs, let me one moment

loumalouomega committed:
Merge branch 'master' into StructMech/displacement_control

loumalouomega committed:
Adding epsilon as constexpr

loumalouomega committed:
Correcting std::abs, and conditions mpore efficient

loumalouomega committed:
Minor style

loumalouomega committed:
Const pointer

loumalouomega committed:
Fix warning

loumalouomega reviewed (dismissed):
OK, I fixed my last comments, green light

zidanmahmoud commented:
Thanks! @loumalouomega

Commit:
fix warning end of non-void function

loumalouomega reviewed (approved):
OK

zidanmahmoud commented:
Just a minor else statement to fix "warning: control reaches end of non-void function", because travis-ci would consider it as an error.

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/displacement_control_with_direction_process.py:
```diff
@@ -0,0 +1,102 @@
+            "model_part_name" : "please_specify_model_part_name",
+            "direction"       : "x, y, or z",
+            "interval"        : [0.0, 1e30],
+            "POINT_LOAD_value"              : 1.0,
```
please use lower case for settings in json

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/displacement_control_with_direction_process.py:
```diff
@@ -0,0 +1,102 @@
+            "direction"       : "x, y, or z",
+            "interval"        : [0.0, 1e30],
+            "POINT_LOAD_value"              : 1.0,
+            "PRESCRIBED_DISPLACEMENT_value" : "please give an expression in terms of the variable x, y, z, t",
```
same

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/displacement_control_with_direction_process.py:
```diff
@@ -0,0 +1,102 @@
+            "interval"        : [0.0, 1e30],
+            "POINT_LOAD_value"              : 1.0,
+            "PRESCRIBED_DISPLACEMENT_value" : "please give an expression in terms of the variable x, y, z, t",
+            "local_axes"      : {}
```
is this being used? if not then I suggest to remove it

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/displacement_control_with_direction_process.py:
```diff
@@ -0,0 +1,102 @@
+        default_settings = KratosMultiphysics.Parameters("""
+        {
+            "help"            : "This process assigns given POINT_LOAD and PRESCRIBED_DISPLACEMENT values (scalar) to the DisplacementControl belonging a certain submodelpart",
+            "mesh_id"         : 0,
```
this is legacy, please remove it from the interface

philbucher commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -0,0 +1,372 @@
+    SizeType mat_size = number_of_nodes * GetBlockSize();
+
+    auto p_load = GetPointLoadInDirection();
+    auto p_disp = GetDisplacementInDirection();
```
const

philbucher commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -0,0 +1,372 @@
+    SizeType number_of_nodes = 1;
+    SizeType mat_size = number_of_nodes * GetBlockSize();
+
+    auto p_load = GetPointLoadInDirection();
```
const

philbucher commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -0,0 +1,372 @@
+/***********************************************************************************/
+/***********************************************************************************/
+
+DisplacementControlCondition::Array1DComponentType* DisplacementControlCondition::GetPointLoadInDirection()
```
this function should be `const`

philbucher commented on applications/StructuralMechanicsApplication/custom_conditions/displacement_control_condition.cpp:
```diff
@@ -0,0 +1,372 @@
+/***********************************************************************************/
+/***********************************************************************************/
+
+DisplacementControlCondition::Array1DComponentType* DisplacementControlCondition::GetDisplacementInDirection()
```
same here

philbucher reviewed (commented):
some minor comments, please make those changes in a small PR

otherwise looks good to me, nice work :+1:
